### ISSUE TITLE
Uninitialized memory read in MAHI code

### ISSUE THREAD

**Issue by fwyzard** [cms-patatrack/cmssw#520]
While validating the recent memory changes, `cuda-memcheck` spotted a read from uninitialised memory in the MAHI code on GPU:
```
CUDA-MEMCHECK
Uninitialized __global__ memory read of size 1
    at 0x000018a0 in src/RecoLocalCalo/HcalRecProducers/src/MahiGPU.cu:359: hcal::mahi::kernel_prep1d_sameNumberOfSamples(...)
    by thread (7,19,0) in block (159,0,0)
```
which points to https://github.com/cms-patatrack/cmssw/blob/9156aad2913ce71da43d4e304e505cda355a3ecb/RecoLocalCalo/HcalRecProducers/src/MahiGPU.cu#L359-L361

A full report can be found for example [here](https://patatrack.web.cern.ch/patatrack/validation/pulls/3dd4b1cc050826346d8527b3ec41c9817ef678b4/RelValTTbar_14TeV-CMSSW_11_1_0_pre8-PU_111X_mcRun3_2021_realistic_v4-v1/development-11634.522-initcheck.out).

**Comment by fwyzard** [comment]
@vkhristenko could you have a look ?

**Comment by fwyzard** [comment]
The error happens on all inputs we test: TTbar, Zmumu, Zee.

A configuration to reproduce the issue is, for example, this one: [step3.py](https://patatrack.web.cern.ch/patatrack/validation/pulls/3dd4b1cc050826346d8527b3ec41c9817ef678b4/RelValTTbar_14TeV-CMSSW_11_1_0_pre8-PU_111X_mcRun3_2021_realistic_v4-v1/development-11634.522-step3.py).

To run on the same TTbar sample from EOS, just drop `file:/gpu_data` from the beginning of the file name.

**Comment by vkhristenko** [comment]
kk

**Comment by vkhristenko** [comment]
the problem is that hcal side was updated for simulation (10 digis for mc vs 8 digis for data) 
https://github.com/cms-sw/cmssw/pull/28948/files

**Comment by fwyzard** [comment]
OK... so, what changes do we need in the GPU code ?

**Comment by vkhristenko** [comment]
Ill prep a pr

**Comment by fwyzard** [comment]
OK, thanks !

So, the problem is only for MC, while data should be fine ?

**Comment by vkhristenko** [comment]
@mariadalfonso 

This is what Maria told me today. For simulation, they added 2 more time samples in the frame. And that is what I saw when debugging using config u pointed to. For data (HLT specifically), should be as currently configured in the GPU version.

**Comment by fwyzard** [comment]
@vkhristenko @mariadalfonso any updates regarding this ?

I would like to build a new release in the next days, and a fix for this would be a welcome feature :-)

**Comment by vkhristenko** [comment]
Monday/Tuesday will create a pr, once I’m back at the office

**Comment by fwyzard** [comment]
Sounds good, thank you very much !

**Comment by vkhristenko** [comment]
1 day delay, sorry!

**Comment by vkhristenko** [comment]
#531 for this, should be validated properly (obv i do that as well...)

**Comment by fwyzard** [comment]
thanks @vkhristenko !

**Comment by fwyzard** [comment]
Fixed by #531 .
